### DISCUSSION TITLE
Adding Heatmaps to Walkways Based on Commuter Density

### DISCUSSION THREAD

**Discussion by the5solae** [mesa/mesa-geo#174] [Q&A]
Hi, I am working on a project to visualize pedestrian traffic patterns in an urban environment. I am currently using a simple line visualization (from a shapefile) to represent the walkways, but I would like to enhance the visualization by incorporating a heatmap that reflects the density of commuters on each walkway. I have been following the **Agents and Network Tutorial**, but I am unsure how to implement this heatmap feature.

**Comment by wang-boyu** [comment]
Hi this is a very interesting use case. One of my colleagues was also wondering similar questions a few weeks ago. The agents-and-networks example currently doesn't have this functionality.

What I can think of at the moment is to first have road network segments as agents, each with its unique id and an attribute for `num_agents`. In the `Commuter` agent, somehow figure out which road segment it is currently on (perhaps with spatial queries such as "within" by using [`get_relation()`](https://github.com/projectmesa/mesa-geo/blob/e10761e30bef509ea270a96e39decc0d97fc1318/mesa_geo/geospace.py#L174C5-L181C67) but this might be slow). Then increment the corresponding road segment agent's `num_agents` field. If you have a large road network then this may not be the most performant way of doing it.

In terms of the visualization you can update the `agent_portrayal` function to assign various colors to road segments according to the `num_agents` values. Alternatively, you may try to do the visualization in jupyter notebook by extracting all agents as geodataframe and plot them using geopandas (a small demo [here](https://github.com/wang-boyu/mesa-examples/blob/gis%2Fsolara-frontend/gis/geo_schelling_points/app.ipynb)).

**Reply by the5solae** [reply]
Thank you for this idea @wang-boyu. I'm not too concerned with speed right now. Need to a PoC first. The walkway agent is already part of the agents in the Agents and Network tutorial, my guess is that I can use that as the road network segments. I'll share any issues I face. Thanks again.

**Reply by the5solae** [reply]
When I try to use any of the queries to see when the commuter intersects/touches/contains/within; is in the neighborhood of any other agent it only returns `None`. 

This is what I did:
I implemented the spacial queries in the `Campus Space` and called it in a commuter method. 

Can you let me know what I'm doing wrong?

Thank you.

**Reply by wang-boyu** [reply]
I'm not entirely sure why it doesn't work, but I'm guessing it may not be easy to perform spatial join of points on lines. Do you wanna try to buffer the lines and do a spatial join of points in polygons instead? But this may return multiple results, since line buffers may overlap.

Another idea is to not use spatial joins at all, and avoid adding road networks as agents. First, when constructing road network, add edge id into the graph (`_nx_graph` in `RoadNetwork` class) as edge attribute. Based on [NetworkX documentation](https://networkx.org/documentation/stable/auto_examples/geospatial/plot_lines.html), information from GeoDataFrame is said to be preserved as edge attribute automatically when using [`momepy.gdf_to_nx()`](https://github.com/projectmesa/mesa-examples/blob/7275b172aad8905584dcffc131d9bbd96992abb7/gis/agents_and_networks/src/space/road_network.py#L22), although I haven't checked it myself.

Then, when we compute agents' paths, instead of returning individual points, try to get individual points plus which line id (i.e., an edge attribute) every point is on. The method to change is: https://github.com/projectmesa/mesa-examples/blob/7275b172aad8905584dcffc131d9bbd96992abb7/gis/agents_and_networks/src/space/utils.py#L60-L76. Given a line or a multiline, it returns a list of points at intervals given by commuter's speed, such as

```python
[Point(0, 0), Point(0.5, 0.5), Point(1, 1)]
```

Instead, you may want something like this

```python
[(Point(0, 0), "LineID_0"), (Point(0.5, 0.5), "LineID_0"), (Point(1, 1), "LineID_1")]
```

Then as each agent moves to the next point, we can easily get which line id the agent is on, without having to do spatial joins. But of course, you may need to update several other methods (such as `RoadNetwork.get_shortest_path()` and `Commuter._redistribute_path_vertices()`) to make them work with it.
